### PR TITLE
Join messages

### LINKED ISSUE #684: Create 'Players Joining' And 'All Players Joined' Information Messages
# For New Contributors
[How to contribute](https://github.com/Bannerlord-Coop-Team/BannerlordCoop/wiki/Getting-Started-as-a-Contributor)

## Description

Add functionality that satisfies the requirement section

## Intended Design

N/A

## Location

Create a new branch based from development.

Create `Coop.Integration` tests for all ICommands used with this new functionality

## Related Issues

N/A

## Requirements

1. When a player starts joining the game, on all clients, send an information message using `SendInformationMessage` that states "A new player is joining the game, pausing"
2. When a player attempts to un-pause while a player is joining, send an information message using `SendInformationMessage` that states "Pausing disabled \<Number of people joining\> players are currently joining the game"
3. When a `PlayerCampaignEntered` event is triggered, If all clients in `IClientRegistry` are not in a [loading state](https://github.com/Bannerlord-Coop-Team/BannerlordCoop/blob/development/source/Coop.Core/Server/Connections/ClientRegistry.cs#L29), from the server, trigger an information message using `SendInformationMessage` that states "All players connected, game can now be un-paused" on all clients
4. When a `PlayerCampaignEntered` event is triggered, If all clients in `IClientRegistry` are not in a [loading state](https://github.com/Bannerlord-Coop-Team/BannerlordCoop/blob/development/source/Coop.Core/Server/Connections/ClientRegistry.cs#L29), from the server, trigger an information message using `SendInformationMessage` that states "All players connected, game can now be un-paused" on the server

## Additional information

N/A

## Definition of Done
- [ ] Class level comments exist for all new classes.
- [ ] XUnit tests exist for every method that does not require the game to be ran.

### PR DESCRIPTION
**Pull request by EgardA**
## PR Checklist

Please check if your PR fulfills the following requirements:

- [ ] All new classes have class-level documentation comments, if there are any at all
- [ ] Tests for the changes have been added (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation update
- [ ] Other... Please describe:

## What is the current behaviour?

Some join messages

## What is the new behaviour?

Resolves #684 

## Other information

### CODE REVIEW

**Comment by garrettluskey** [comment]
Currently tests are failing and requirement 2 in #684 is not covered as far as I can tell

**Comment by EgardA** [comment]
You are correct, fixing.

File: source/Coop.Core/Client/States/MainMenuState.cs [hunk 1]
```diff
@@ -35,6 +36,7 @@ public override void Connect()
     internal void Handle_NetworkConnected(MessagePayload<NetworkConnected> obj)
     {
         Logic.ValidateModules();
+        messageBroker.Publish(this, new SendInformationMessage("A new player is joining the game, pausing"));
```
Comment (garrettluskey) [inline source/Coop.Core/Client/States/MainMenuState.cs]: Instead create a JoinMessage service that subscribes to the same events. This is separate functionality than the states here

**Review by garrettluskey — CHANGES_REQUESTED** [review]

File: source/Coop.Core/Server/Connections/ClientRegistry.cs [hunk 2]
```diff
@@ -61,6 +61,8 @@ internal void PlayerJoiningHandler(MessagePayload<PlayerConnected> obj)
         var playerPeer = obj.What.PlayerPeer;
         var connectionLogic = connectionLogicFactory.CreateLogic(playerPeer);
         ConnectionStates.Add(playerPeer, connectionLogic);
+
+        messageBroker.Publish(this, new SendInformationMessage("A new player is joining the game, pausing"));
```
Comment (garrettluskey) [inline source/Coop.Core/Server/Connections/ClientRegistry.cs]: Separate this into a separate service as well

File: source/Coop.Tests/Server/Connections/States/LoadingStateTests.cs [hunk 3]
```diff
@@ -79,7 +79,7 @@ public void PlayerCampaignEntered_ValidPlayerId()
 
 
             // Assert
-            Assert.Single(MockMessageBroker.PublishedMessages);
+            Assert.True(MockMessageBroker.PublishedMessages.Count == 2);
```
Comment (garrettluskey) [inline source/Coop.Tests/Server/Connections/States/LoadingStateTests.cs]: Revert this please

File: source/Coop.Core/Server/Services/Time/Handlers/TimeHandler.cs [hunk 4]
```diff
@@ -39,7 +40,13 @@ public void Dispose()
...
     internal void Handle_NetworkRequestTimeSpeedChange(MessagePayload<NetworkRequestTimeSpeedChange> obj)
     {
-        if (AnyLoaders()) return;
+        if (AnyLoaders())
+        {
+            int loadingPeers = _clientRegistry.LoadingPeers.Count;
+            _messageBroker.Publish(this, new SendInformationMessage("Pausing disabled, " + loadingPeers + " player(s) are currently joining the game"));
+            _network.SendAll(new SendInformationMessage("Pausing disabled, " + loadingPeers + " player(s) are currently joining the game"));
+            return;
+        }
```
Comment (garrettluskey) [inline source/Coop.Core/Server/Services/Time/Handlers/TimeHandler.cs]: These should be in the same class

File: source/Coop.Core/Server/Services/Time/Handlers/TimeHandler.cs [hunk 5]
```diff
@@ -48,7 +55,13 @@ internal void Handle_NetworkRequestTimeSpeedChange(MessagePayload<NetworkRequest
...
     internal void Handle_TimeSpeedChanged(MessagePayload<AttemptedTimeSpeedChanged> obj)
     {
-        if (AnyLoaders()) return;
+        if (AnyLoaders())
+        {
+            int loadingPeers = _clientRegistry.LoadingPeers.Count;
+            _messageBroker.Publish(this, new SendInformationMessage("Pausing disabled, " + loadingPeers + " player(s) are currently joining the game"));
+            _network.SendAll(new SendInformationMessage("Pausing disabled, " + loadingPeers + " player(s) are currently joining the game"));
+            return;
+        }
```
Comment (garrettluskey) [inline source/Coop.Core/Server/Services/Time/Handlers/TimeHandler.cs]: These should be in the same class

File: source/Coop.Core/Server/Services/Players/Handlers/PlayerConnectionServerMessageHandler.cs [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+
+        private void PlayerCampaignEnteredHandler(MessagePayload<NetworkPlayerCampaignEntered> obj)
+        {
+            var playerId = (NetPeer)obj.Who;
+
+            if (!clientRegistry.PlayersLoading)
+            {
+                messageBroker.Publish(this, new SendInformationMessage("All players connected, game can now be un-paused"));
+                network.SendAllBut(playerId, new SendInformationMessage("All players connected, game can now be un-paused"));
+            }
```
Comment (garrettluskey) [inline source/Coop.Core/Server/Services/Players/Handlers/PlayerConnectionServerMessageHandler.cs]: These should be in the same class

File: source/Coop.Core/Server/Connections/States/LoadingState.cs [hunk 7]
```diff
@@ -10,11 +12,19 @@ namespace Coop.Core.Server.Connections.States;
 public class LoadingState : ConnectionStateBase
 {
     private readonly IMessageBroker messageBroker;
+    private readonly IClientRegistry clientRegistry;
+    private readonly INetwork network;
```
Comment (garrettluskey) [inline source/Coop.Core/Server/Connections/States/LoadingState.cs]: These likely aren't used anymore

**Review by garrettluskey — CHANGES_REQUESTED** [review]

**Comment by garrettluskey** [comment]
All of the logging should exist in the same class. Having all the functionality in a single place makes it easier to maintain and not add lines to unrelated classes

File: source/Coop.Core/Common/Services/Connection/Handlers/PlayerConnectionServerMessageHandler.cs [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+            messageBroker.Publish(this, new SendInformationMessage("A new player is joining the game, pausing"));
+        }
+        private void NetworkRequestTimeSpeedChange(MessagePayload<NetworkRequestTimeSpeedChange> obj)
+        {
+            if (AnyLoaders())
+            {
+                int loadingPeers = clientRegistry.LoadingPeers.Count;
+                messageBroker.Publish(this, new SendInformationMessage("Pausing disabled, " + loadingPeers + " player(s) are currently joining the game"));
+                network.SendAll(new SendInformationMessage("Pausing disabled, " + loadingPeers + " player(s) are currently joining the game"));
+            }
```
Comment (garrettluskey) [inline source/Coop.Core/Common/Services/Connection/Handlers/PlayerConnectionServerMessageHandler.cs]: this logic only works on the server

File: source/Coop.Core/Common/Services/Connection/Handlers/PlayerConnectionServerMessageHandler.cs [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+
+        private void PlayerCampaignEnteredHandler(MessagePayload<NetworkPlayerCampaignEntered> obj)
+        {
+            var playerId = (NetPeer)obj.Who;
+
+            if (!clientRegistry.PlayersLoading)
+            {
+                messageBroker.Publish(this, new SendInformationMessage("All players connected, game can now be un-paused"));
+                network.SendAllBut(playerId, new SendInformationMessage("All players connected, game can now be un-paused"));
+            }
```
Comment (garrettluskey) [inline source/Coop.Core/Common/Services/Connection/Handlers/PlayerConnectionServerMessageHandler.cs]: this logic only works on the server

File: source/Coop.Core/Common/Services/Connection/Handlers/PlayerConnectionServerMessageHandler.cs [hunk 10]
```diff
@@ -0,0 +1,76 @@
...
+using Coop.Core.Server.Connections.Messages;
+using Coop.Core.Server.Services.Time.Messages;
+using GameInterface.Services.GameDebug.Messages;
+using LiteNetLib;
+using Serilog;
+using System.Linq;
+
+namespace Coop.Core.Common.Services.Connection.Handlers
+{
+    public class PlayerConnectionServerMessageHandler
```
Comment (garrettluskey) [inline source/Coop.Core/Common/Services/Connection/Handlers/PlayerConnectionServerMessageHandler.cs]: This class is never initialized, move to the server directory and have this class inherit IHandler to be collected by https://github.com/Bannerlord-Coop-Team/BannerlordCoop/blob/development/source/Coop.Core/Server/ServerModule.cs#L34

File: source/Coop.Core/Common/Services/Connection/Handlers/PlayerConnectionServerMessageHandler.cs [hunk 11]
```diff
@@ -0,0 +1,76 @@
...
+
+            messageBroker.Subscribe<NetworkPlayerCampaignEntered>(PlayerCampaignEnteredHandler);
+            messageBroker.Subscribe<PlayerConnected>(PlayerConnectedHandler);
+            messageBroker.Subscribe<NetworkRequestTimeSpeedChange>(NetworkRequestTimeSpeedChange);
+            messageBroker.Subscribe<NetworkConnected>(Handle_NetworkConnected);
+        }
+
+        internal void Handle_NetworkConnected(MessagePayload<NetworkConnected> obj)
+        {
+            messageBroker.Publish(this, new SendInformationMessage("A new player is joining the game, pausing"));
```
Comment (garrettluskey) [inline source/Coop.Core/Common/Services/Connection/Handlers/PlayerConnectionServerMessageHandler.cs]: The game does not pause exactly when a player connects, the game instead pauses [here](https://github.com/Bannerlord-Coop-Team/BannerlordCoop/blob/development/source/Coop.Core/Server/Connections/States/TransferSaveState.cs#L27)

File: source/Coop.Core/Common/Services/Connection/Handlers/PlayerConnectionServerMessageHandler.cs [hunk 12]
```diff
@@ -0,0 +1,76 @@
...
+        }
+
+        private void PlayerCampaignEnteredHandler(MessagePayload<NetworkPlayerCampaignEntered> obj)
+        {
+            var playerId = (NetPeer)obj.Who;
+
+            if (!clientRegistry.PlayersLoading)
+            {
+                messageBroker.Publish(this, new SendInformationMessage("All players connected, game can now be un-paused"));
```
Comment (garrettluskey) [inline source/Coop.Core/Common/Services/Connection/Handlers/PlayerConnectionServerMessageHandler.cs]: 2x of same string, separate into variable

File: source/Coop.Core/Common/Services/Connection/Handlers/PlayerConnectionServerMessageHandler.cs [hunk 13]
```diff
@@ -0,0 +1,76 @@
...
+        private void PlayerConnectedHandler(MessagePayload<PlayerConnected> obj)
+        {
+            messageBroker.Publish(this, new SendInformationMessage("A new player is joining the game, pausing"));
+        }
+        private void NetworkRequestTimeSpeedChange(MessagePayload<NetworkRequestTimeSpeedChange> obj)
+        {
+            if (AnyLoaders())
+            {
+                int loadingPeers = clientRegistry.LoadingPeers.Count;
+                messageBroker.Publish(this, new SendInformationMessage("Pausing disabled, " + loadingPeers + " player(s) are currently joining the game"));
```
Comment (garrettluskey) [inline source/Coop.Core/Common/Services/Connection/Handlers/PlayerConnectionServerMessageHandler.cs]: 2x of same string, separate into variable

**Review by garrettluskey — CHANGES_REQUESTED** [review]

File: source/Coop.Core/Coop.Core.csproj [hunk 14]
```diff
@@ -76,4 +76,7 @@
 		  <HintPath>..\..\mb2\bin\Win64_Shipping_Client\TaleWorlds.Library.dll</HintPath>
 	  </Reference>
   </ItemGroup>
+  <ItemGroup>
+    <Folder Include="Common\Services\Connection\Handlers\" />
+  </ItemGroup>
```
Comment (garrettluskey) [inline source/Coop.Core/Coop.Core.csproj]: Could you remove this? this will keep that unused folder in the project

**Review by garrettluskey — CHANGES_REQUESTED** [review]

**Pull request merged**